[Quant][FX] Lower QConvAdd2d for onednn backend

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91155
* #91154
* __->__ #91153
* #91152

**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

leslie-fang-intel committed:
[Quant][FX] Lower QConvAdd2d for onednn backend

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -937,6 +1062,9 @@ def _lower_to_native_backend(
     operator signature so they can be lowered with the same function
     """
     _lower_static_weighted_ref_module(model, qconfig_map)
+    if torch.backends.quantized.engine == 'onednn':
+        # Currently, this pass only works for onednn with conv2d_add and conv2d_add_relu fusion
+        _lower_static_weighted_ref_module_with_two_dq_inputs(model, qconfig_map)
```
This rewrite function sounds general to me, not needed to add a special `onednn` backend check here?

leslie-fang-intel replied:
Thanks for comment, remove the checks here.

jgong5 reviewed (approved):
LGTM

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on torch/ao/ns/fx/mappings.py:
```diff
@@ -373,6 +373,7 @@ def get_base_name_to_sets_of_related_ops() -> Dict[str, Set[NSNodeTargetType]]:
 
     for source_to_double_target in (
         _lower_to_native_backend.STATIC_LOWER_FUSED_MODULE_MAP,
+        _lower_to_native_backend.STATIC_LOWER_FUSED_MODULE_TWO_DQ_NODES_MAP,
```
nit: it might be clearer to rename `TWO_DQ_NODES` to `TWO_INPUTS` I think

jerryzh168 commented on torch/ao/quantization/backend_config/onednn.py:
```diff
@@ -103,10 +105,162 @@ def _fuse_linear_bn_leaky_relu(is_qat, linear, bn, leaky_relu):
+
+def _fuse_conv_bn_add_left(is_qat, add, bn_conv, _):
+    bn, conv = bn_conv
+    fused_conv = nn.utils.fusion.fuse_conv_bn_eval(conv, bn)
```
wait, what about QAT?

leslie-fang-intel replied:
Thanks for the comment. Add a check for the QAT cases.

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -470,6 +479,62 @@ def _match_static_pattern(
 
     return (q_node, relu_node, ref_node)
 
+def _match_static_pattern_with_two_dq_inputs(
```
nit: two_dq_inputs -> two_inputs

leslie-fang-intel replied:
Thanks for the suggestions and changed.

jerryzh168 commented on torch/ao/quantization/backend_config/onednn.py:
```diff
@@ -103,10 +105,162 @@ def _fuse_linear_bn_leaky_relu(is_qat, linear, bn, leaky_relu):
+#   \   /
+#    add
+
+def _fuse_conv_add(is_qat, add, _, conv):
```
nit: should we name this `_fuse_conv_bn_add_right` to make it clearer

leslie-fang-intel replied:
Thanks for the comments. Rename the relevant functions with suffix of  `_right` to make code clearer.

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@jerryzh168 I have changed according to your comments. Could you help to take a look of this PR again?

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168 Is there any other comments to this PR? Could you help to take a look again?

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168, Could you also take a review of this PR again?

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -396,14 +396,16 @@ def _test_conv_api_impl(
-            # workaround in this PR to return from here, since the below lowering part enabled in next PR
-            # We will enable below check in next PR
-            return
+        class _FusedModule_two_input_args(torch.nn.intrinsic._FusedModule):
```
do we need this?

leslie-fang-intel replied:
Yes, I think so. That's because original `torch.nn.intrinsic._FusedModule` only support one input.

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -447,6 +448,114 @@ def test_linear_tanh_not_fused_by_default(self):
+
+        options = itertools.product(
+            [True, False],  # with_bn
+            [False],  # with_relu
```
is relu supported?

leslie-fang-intel replied:
Oh, yes. The support of relu is added in the following up 2 PRs.

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -447,6 +448,114 @@ def test_linear_tanh_not_fused_by_default(self):
+    def test_fuse_conv_bn_add_relu_by_default(self):
+        options = itertools.product(
+            [True, False],  # with_bn
+            [False],  # with_relu
```
same question for relu

leslie-fang-intel replied:
The support of relu has been added in the following up PRs.

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -268,6 +268,15 @@ def should_skip_lowering(op: torch.fx.node.Node, qconfig_map: Dict[str, QConfigA
 }
 
+# The difference between STATIC_LOWER_FUSED_MODULE_TWO_INPUTS_MAP and STATIC_LOWER_FUSED_MODULE_MAP:
+# The refer node inside STATIC_LOWER_FUSED_MODULE_TWO_INPUTS_MAP has 2 dq input nodes.
```
`dq input nodes` --> `inputs`

leslie-fang-intel replied:
Thanks for the comments and changed.

jerryzh168 reviewed (approved):
lgtm, thanks!

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][FX] Lower QConvAdd2d for onednn backend"


**Summary**
Add quantization mappings for QConvAdd2d for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
```
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_onednn
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_by_default
python -m pytest test_quantization.py -k test_fuse_conv_bn_add_relu_lowering
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
